spec/consensus/proposer-based-timestamp/pbts-algorithm.md

```diff
@@ -0,0 +1,167 @@
+The proposal time is considered `timely` by `p` when:
+
+1. `now_p >= v.time - PRECISION`
+1. `now_p <= v.time + MSGDELAY + PRECISION`
+
+The first condition derives from the fact that the generation and sending of `v` precedes its reception.
+The minimum receiving time `now_p` for `v.time` be considered `timely` by `p` is derived from the extreme scenario when
+the clock of `p` is `PRECISION` *behind* of the clock of the proposer of `v`, and the proposal's transmission delay is `0` (minimum).
+
+The second condition derives from the assumption of an upper bound for the transmission delay of a proposal.
+The maximum receiving time `now_p` for `v.time` be considered `timely` by `p` is derived from the extreme scenario when
+the clock of `p` is `PRECISION` *ahead* of the clock of the proposer of `v`, and the proposal's transmission delay is `MSGDELAY` (maximum).
+
+## Updated Consensus Algorithm
+
+The following changes are proposed for the algorithm in the [arXiv paper][arXiv].
+
+### Updated `StartRound` function
+
+There are two additions to operation of the **proposer** of a round:
+
+1. to ensure time monotonicity, the proposer does not propose a value until its
+current local time, represented by `now_p` in the algorithm, 
+becomes greater than the previously decided value's time
+1. when the proposer produce a new proposal it sets the proposal's time to its current local time
+   - no changes are made to the logic when a proposer has a non-nil `validValue_p`, which retains its original proposal time.
+
+#### **[PBTS-ALG-STARTROUND.1]**
+
+```go
+function StartRound(round) {
+   round_p ← round
+   step_p ← propose
+   if proposer(h_p, round_p) = p {
+      if validValue_p != nil {
+         proposal ← validValue_p // proposal.time unchanged
+      } else {
+         wait until now_p > decision_p[h_p-1].time // time monotonicity
+         proposal ← getValue()
+         proposal.time ← now_p // proposal time
+      }
+      broadcast ⟨PROPOSAL, h_p, round_p, proposal, validRound_p⟩
+   } else {
+      schedule OnTimeoutPropose(h_p,round_p) to be executed after timeoutPropose(round_p)
+   }
+}
+```
+
+### Updated upon clause of line 22
+
+The rule on line 22 applies to values `v` proposed for the first time, i.e.,
+for proposals not backed by `2f + 1 PREVOTE`s for `v` in a previous round.
+The `PROPOSAL` message, in this case, has a valid round `vr = -1`.
+
+The new rule for issuing a `PREVOTE` for a proposed value `v`
+requires the proposal time to be `timely`:
+
+#### **[PBTS-ALG-UPON-PROP.1]**
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v, −1⟩ from proposer(h_p, round_p) while step_p = propose do {
+   if timely(v.time) ∧ valid(v) ∧ (lockedRound_p = −1 ∨ lockedValue_p = v) {
+      broadcast ⟨PREVOTE, h_p, round_p, id(v)⟩ 
+   }
+   else {
+      broadcast ⟨PREVOTE, h_p, round_p, nil⟩ 
+   }
+   step_p ← prevote
+}
+```
+
+The `timely` predicate considers the time at which the `PROPOSAL` message is received.
```

Comment:
This observation reflect the fact that we might receive a `PROPOSAL` before we entered the round or are at the propose round step. This also aligns with the implementation, which sets the receive time of a `PROPOSAL`, instead of considering its processing time.